sdk/go/tappd/client.go

```diff
@@ -0,0 +1,274 @@
+	client := &TappdClient{
+		endpoint:   "",
+		baseURL:    "",
+		httpClient: &http.Client{},
+		logger:     slog.Default(),
+	}
+
+	for _, opt := range opts {
+		opt(client)
+	}
+
+	client.endpoint = client.getEndpoint()
+
+	if strings.HasPrefix(client.endpoint, "http://") || strings.HasPrefix(client.endpoint, "https://") {
+		client.baseURL = client.endpoint
+	} else {
+		client.baseURL = "http://localhost"
+		client.httpClient = &http.Client{
+			Transport: &http.Transport{
+				DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+					return net.Dial("unix", client.endpoint)
+				},
+			},
+		}
+	}
+
+	return client
+}
+
+// Returns the appropriate endpoint based on environment and input. If the
+// endpoint is empty, it will use the simulator endpoint if it is set in the
+// environment through DSTACK_SIMULATOR_ENDPOINT. Otherwise, it will use the
+// default endpoint at /var/run/tappd.sock.
+func (c *TappdClient) getEndpoint() string {
+	if c.endpoint != "" {
+		return c.endpoint
+	}
+	if simEndpoint, exists := os.LookupEnv("DSTACK_SIMULATOR_ENDPOINT"); exists {
+		c.logger.Info("using simulator endpoint", "endpoint", simEndpoint)
+		return simEndpoint
+	}
+	return "/var/run/tappd.sock"
+}
+
+// Sends an RPC request to the Tappd service.
+func (c *TappdClient) sendRPCRequest(ctx context.Context, path string, payload interface{}) ([]byte, error) {
+	jsonData, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", c.baseURL+path, bytes.NewBuffer(jsonData))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	return io.ReadAll(resp.Body)
+}
+
+// Derives a key from the Tappd service. If altNames is empty or nill, it will
+// not be included inthe request.
+func (c *TappdClient) DeriveKey(ctx context.Context, path string, subject string, altNames []string) (*DeriveKeyResponse, error) {
+	payload := map[string]interface{}{
+		"path":    path,
+		"subject": subject,
+	}
+	if len(altNames) > 0 {
+		payload["alt_names"] = altNames
+	}
+
+	data, err := c.sendRPCRequest(ctx, "/prpc/Tappd.DeriveKey", payload)
+	if err != nil {
+		return nil, err
+	}
+
+	var response DeriveKeyResponse
+	if err := json.Unmarshal(data, &response); err != nil {
+		return nil, err
+	}
+	return &response, nil
+}
+
+// Sends a TDX quote request to the Tappd service.
+func (c *TappdClient) TdxQuote(ctx context.Context, reportData []byte, hashAlgorithm QuoteHashAlgorithm) (*TdxQuoteResponse, error) {
+	hexData := hex.EncodeToString(reportData)
+	if hashAlgorithm == RAW {
+		if len(hexData) > 128 {
+			return nil, fmt.Errorf("report data is too large, it should be less than 128 characters when hash_algorithm is raw")
```

Comment:
```suggestion
			return nil, fmt.Errorf("report data is too large, it should be at most 64 bytes when hashAlgorithm is RAW")
```

Reply:
@kvinwang It looks like he's copying from the original implementation (refs [1](https://github.com/Dstack-TEE/dstack/blob/master/sdk/js/src/index.ts#L232), [2](https://github.com/Dstack-TEE/dstack/blob/master/sdk/python/src/dstack_sdk/tappd_client.py#L137)). Should we consistently specify the data size before or after hex encoding?

Reply:
> @kvinwang It looks like he's copying from the original implementation (refs [1](https://github.com/Dstack-TEE/dstack/blob/master/sdk/js/src/index.ts#L232), [2](https://github.com/Dstack-TEE/dstack/blob/master/sdk/python/src/dstack_sdk/tappd_client.py#L137)). Should we consistently specify the data size before or after hex encoding?

I figured it out because the argument the caller passed in was bytes, so the error reported to the caller should be a hint to the caller about what's wrong with the argument.

Reply:
I agree! This is from the other implementations, but since in this case we're enforcing a bytes representation, it makes more sense to refer to the bytes length. As far as I understand there's no need to standardize errors here

Reply:
Actually, now that I see the Python implementation again, I do think it'd be better to be consistent on the bytes data size - if the user passes a string, this is not a hex string but a string to be UTF-8 encoded. It may be more understandable to the user to relate his string to the byte size than the hex encoding length.